rasa/shared/core/domain.py

```diff
@@ -256,61 +259,127 @@ def merge(self, domain: Optional["Domain"], override: bool = False) -> "Domain":
-            tempDict2: Dict[Text, Any],
-            override_existing_values: bool = False,
-        ) -> Dict[Text, Any]:
+        if override:
+            config = domain_dict["config"]
+            for key, val in config.items():
+                combined["config"][key] = val
+
+        if override or self.session_config == SessionConfig.default():
+            combined[SESSION_CONFIG_KEY] = domain_dict[SESSION_CONFIG_KEY]
 
-            if override_existing_values:
-                merge_dicts, b = tempDict1.copy(), tempDict2.copy()
+        if combined[KEY_INTENTS] or domain_dict[KEY_INTENTS]:
+            combined[KEY_INTENTS] = self.merge_lists_of_dicts(
+                combined[KEY_INTENTS], domain_dict[KEY_INTENTS], override
+            )
 
-            else:
-                merge_dicts, b = tempDict2.copy(), tempDict1.copy()
+        # remove existing forms from new actions
+        for form in combined[KEY_FORMS]:
+            if form in domain_dict[KEY_ACTIONS]:
+                domain_dict[KEY_ACTIONS].remove(form)
+
+        for key in [KEY_ENTITIES, KEY_ACTIONS, KEY_E2E_ACTIONS]:
+            combined[key] = self.merge_lists(combined[key], domain_dict[key])
 
-            merge_dicts.update(b)
-            return merge_dicts
+        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
+            combined[key] = self.merge_dicts(combined[key], domain_dict[key], override)
 
-        def merge_lists(list1: List[Any], list2: List[Any]) -> List[Any]:
-            return sorted(list(set(list1 + list2)))
+        return self.__class__.from_dict(combined)
 
-        def merge_lists_of_dicts(
-            dict_list1: List[Dict],
-            dict_list2: List[Dict],
-            override_existing_values: bool = False,
-        ) -> List[Dict]:
-            dict1 = {list(i.keys())[0]: i for i in dict_list1}
-            dict2 = {list(i.keys())[0]: i for i in dict_list2}
-            merged_dicts = merge_dicts(dict1, dict2, override_existing_values)
-            return list(merged_dicts.values())
+    def merge_domain_dicts(
+        self, domain1: Dict, domain2: Dict, override: bool = False
+    ) -> Dict[Text, Any]:
+        """Merges this domain dict with another one, combining their attributes.
+
+        This is used when multiple domain yml files are configured in a single
+        directory. Unlike the merge method above, which merges Domain objects by
+        creating each object then merging it with the previous, this method merges
+        domain dicts, and ensures all attributes (like ``intents``, ``entities``, and
+        ``actions``) are known to the Domain when the object is created.
+
+        List attributes like ``intents`` and ``actions`` are deduped
+        and merged. Single attributes are taken from `domain1` unless
+        override is `True`, in which case they are taken from `domain2`.
+        """
+        if not domain2:
+            return domain1
+
+        if domain1 == {}:
```

Comment:
Since empty dictionaries evaluate to `False` then maybe it's better to do it as it's done above? 
I mean `if not domain1` instead of `if domain == {}`